File: tests/SE2_SLAM/test_Odometry.cpp
#include "gtest/gtest.h"
#include <Eigen/Dense>
#include <InEKF/Core>
#include <InEKF/SE2_SLAM>

bool MatrixEquality(const Eigen::MatrixXd &lhs, const Eigen::MatrixXd &rhs) {
  return lhs.isApprox(rhs, 1e-6);
}

TEST(OdometryProcess, f){
    InEKF::SE2<> state;
    InEKF::SE2<> U(.1, 1, 2);

    InEKF::OdometryProcess op;

    EXPECT_PRED2(MatrixEquality, op.f(U, 1, state)(), U());
}

TEST(OdometryProcess, MakePhi){
    InEKF::SE2<> state;
    InEKF::SE2<> U(.1, 1, 2);

    InEKF::OdometryProcess op;
    Eigen::Matrix3d Phi;

    // check right
    Phi = op.MakePhi(U, 1, state, InEKF::RIGHT);
    EXPECT_PRED2(MatrixEquality, Phi, Eigen::Matrix3d::Identity());

    // check left
    Phi = op.MakePhi(U, 1, state, InEKF::LEFT);
    EXPECT_PRED2(MatrixEquality, Phi, U.Ad());
}

File: python/tests/SE2_SLAM/test_Odometry.py
import numpy as np
from numpy.testing import assert_allclose
from inekf import SE2, OdometryProcess, ERROR
import pytest

def test_f():
    state = SE2()
    u = SE2(.1, 1, 2)

    op = OdometryProcess()

    assert_allclose(op.f(u, 1, state).mat, u.mat)

def test_makePhi():
    state = SE2()
    u = SE2(.1, 1, 2)

    op = OdometryProcess()

    phi = op.makePhi(u, 1, state, ERROR.RIGHT)
    assert_allclose(phi, np.eye(3))

    phi = op.makePhi(u, 1, state, ERROR.LEFT)
    assert_allclose(phi, u.inverse.Ad)